### PR TITLE
Cleanup ATEN Dropout Interface.

### PR DESCRIPTION
**Pull request by kevinstephano**

### CODE REVIEW

**Comment by kevinstephano** [comment]
Ready to be reviewed!

File: aten/src/ATen/native/cuda/Dropout.cu [hunk 1]
```diff
@@ -312,10 +317,11 @@ inline void launcher(
 } //anonymous namespace
 
 std::tuple<Tensor,Tensor>
-fused_dropout_cuda(const Tensor& self, double p, c10::optional<Generator> gen_){
-  auto gen = get_generator_or_default<CUDAGeneratorImpl>(gen_, cuda::detail::getDefaultCUDAGenerator());
+native_dropout_cuda(const Tensor& self, double p, double scale, bool train){
```
Comment (jjsjann123) [inline aten/src/ATen/native/cuda/Dropout.cu:335]: Note to myself: updated signature.

File: aten/src/ATen/core/NamedRegistrations.cpp [hunk 2]
```diff
@@ -12,7 +12,7 @@ TORCH_LIBRARY_IMPL(aten, Named, m) {
   m.impl("_bmm", CppFunction::makeFallthrough());
   m.impl("_bmm.out", CppFunction::makeFallthrough());
   m.impl("_cdist_forward", CppFunction::makeFallthrough());
-  m.impl("_fused_dropout", CppFunction::makeFallthrough());
+  m.impl("native_dropout", CppFunction::makeFallthrough());
```
Comment (jjsjann123) [inline aten/src/ATen/core/NamedRegistrations.cpp:15]: Note to myself: `_fused_dropout` was tested in test_cuda.py

Comment (jjsjann123) [reply aten/src/ATen/core/NamedRegistrations.cpp:15]: https://github.com/pytorch/pytorch/blob/6e3e57095cab409e5c346d57ee42fd89c0584708/test/test_cuda.py#L2936
https://github.com/pytorch/pytorch/blob/c458558334e4aef5107bb60807c76ea9a7078727/tools/code_analyzer/default_op_deps.yaml#L1048

File: aten/src/ATen/native/Dropout.cpp [hunk 3]
```diff
@@ -82,13 +83,37 @@ ALIAS_SPECIALIZATION(_feature_alpha_dropout, true,  true )
...
+
 Tensor dropout(const Tensor& input, double p, bool train) {
+  TORCH_CHECK(p >= 0 && p <= 1, "dropout probability has to be between 0 and 1, but got ", p);
   auto result = [&]() {
     NoNamesGuard guard;
-    if (train && is_fused_kernel_acceptable(input, p)) {
-      return std::get<0>(at::_fused_dropout(input, 1 - p));
+    double p1m = 1. - p;
+    // Check for probability of zero to avoid divide by zero and NaN results
+    double scale = p1m == 0 ? 0. : 1. / p1m;
```
Comment (jjsjann123) [inline aten/src/ATen/native/Dropout.cpp:111]: nitpick. We are removing the short-cut and launching no-op kernels for special case p1m (0.0 & 1.0)

File: tools/autograd/derivatives.yaml [hunk 4]
```diff
@@ -421,8 +421,8 @@
   self: handle_r_to_c(self.scalar_type(), grad.conj() * other)
   other: handle_r_to_c(other.scalar_type(), grad * self)
 
-- name: _fused_dropout(Tensor self, float p, Generator? generator=None) -> (Tensor, Tensor)
-  self: _fused_dropout_backward(grad, result1, p)
+- name: native_dropout(Tensor input, float p, float scale, bool train) -> (Tensor, Tensor)
```
Comment (jjsjann123) [inline tools/autograd/derivatives.yaml:424]: note to myself: gradgrad in JIT is broken. We need to add an entry for native_dropout_backward here.

File: test/test_jit_cuda_fuser.py [hunk 5]
```diff
@@ -1640,6 +1657,73 @@ def test2(x: torch.Tensor, y: torch.Tensor):
...
+        grads = torch.randn([10, 4, 8], dtype=dtype, device=device)
+
+        def t(x: torch.Tensor, p: float, train: bool):
+            o = torch.nn.functional.dropout(x, p, training=train)
+            o = o + 1.0
+            return o
+
+        t_jit = torch.jit.script(t)
+
+        # The drop probability needs to be set to zero given that the order of picking random
```
Comment (jjsjann123) [inline test/test_jit_cuda_fuser.py:1743]: I thought for small tensor, our launch config should be giving identical RNG numbers (if we disable loop switching).

File: test/test_jit_cuda_fuser.py [hunk 6]
```diff
@@ -1640,6 +1657,73 @@ def test2(x: torch.Tensor, y: torch.Tensor):
...
+        # numbers between eager mode and the jit is different
+        self._run_training_helper(t_jit, t, grads, x, 0.0, True)
+
+    @unittest.skipIf(not RUN_CUDA, "requires CUDA")
+    @unittest.skipIf(GRAPH_EXECUTOR != ProfilingMode.PROFILING,
+                     "Requires fusion optimization pass to be effective")
+    def test_dropout_training_prob_check(self):
```
Comment (jjsjann123) [inline test/test_jit_cuda_fuser.py:1750]: Nice!

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 7]
```diff
@@ -726,13 +726,8 @@ Tensor repeat_backward(Tensor grad, IntArrayRef repeats, IntArrayRef input_shape
 }
 
 // p1m == 1 - p
```
Comment (jjsjann123) [inline torch/csrc/autograd/FunctionsManual.cpp]: comment should be updated.

File: torch/csrc/jit/codegen/cuda/graph_fuser.cpp [hunk 8]
```diff
@@ -839,7 +839,7 @@ struct CudaGraphFuser {
           fmap(tensor_inputs, [&](Value* v) { return shape_of.at(v); });
       AT_ASSERT(!shapes.empty());
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/graph_fuser.cpp:840]: Note to myself:
We should assert/throw a warning here on (n->outputs().size() == 1). Meanwhile, add a special rule for dropout.

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 9]
```diff
@@ -720,7 +780,6 @@ class IrParser {
             return OperatorType::Normalization;
           });
     }
-
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/parser.cpp]: is this clang-format?

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 10]
```diff
@@ -1473,6 +1532,26 @@ bool insertProfileIValue(ProfilingRecord* pr, Node* node, size_t offset) {
...
+          "aten::dropout(Tensor input, float p, bool train) -> Tensor")
+          ->schema();
+  static auto native_dropout_schema =
+      getOperatorForLiteral(
+          "aten::native_dropout(Tensor input, float p, float scale, bool train) -> (Tensor, Tensor)")
+          ->schema();
+  if (node->matches(dropout_schema) || node->matches(native_dropout_schema)) {
+    switch (offset) {
+      // argument 2: Is training?
+      case 2:
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/parser.cpp]: [BUG] the offset here is inconsistent between `native_dropout` (3) and `dropout` (2)

File: aten/src/ATen/native/Dropout.cpp (hunk 3 above)
Comment (kevinstephano) [reply aten/src/ATen/native/Dropout.cpp:111]: I actually preferred removing the shortcuts.  I thought it makes things simpler without conditionals.  Someone who is adding the dropout operator is not using it for cases 0. and 1., except for debug use with case 1. to remove randomness, therefore, special casing them for performance did not make sense to me beyond making sure that case 0. does not cause a divide-by-zero fault.

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 11]
```diff
@@ -516,6 +516,66 @@ class IrParser {
...
+      registerParseRule(
+          ptr_op,
+          [](const Node* node,
+             std::unordered_map<size_t, CgValue>& value_map) -> void {
+            auto input = value_map[node->input(0)->unique()];
+            auto prob = value_map[node->input(1)->unique()];
+            auto scale = value_map[node->input(2)->unique()];
+            auto train = constant_as<bool>(node->input(3));
+
+            TORCH_INTERNAL_ASSERT(train.has_value() and train.value(),
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/parser.cpp]: Wasn't dropout train==False just a tensor forwarding?
It does introduce alias (which breaks the schema), but we can just forward the tensor as-is.

Checkout `_grad_sum_to_size` where `!size_to->empty()`

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 12]
```diff
@@ -516,6 +516,66 @@ class IrParser {
...
+      auto ptr_op = getOperatorForLiteral(
+          "aten::dropout(Tensor input, float p, bool train) -> Tensor");
+      registerParseRule(
+          ptr_op,
+          [](const Node* node,
+             std::unordered_map<size_t, CgValue>& value_map) -> void {
+            auto input = value_map[node->input(0)->unique()];
+            auto train = constant_as<bool>(node->input(2));
+
+            TORCH_INTERNAL_ASSERT(train.has_value() && train.value() == false,
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/parser.cpp]: ditto.

**Review by jjsjann123 — CHANGES_REQUESTED** [review]
A tiny bug in profiling_ivalue.
Other than that, a few minor comments, but we can merge the PR and iron them out in following PRs.

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 10 above)
Comment (kevinstephano) [reply torch/csrc/jit/codegen/cuda/parser.cpp]: Nice catch, I forgot I changed the signature.

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 11 above)
Comment (kevinstephano) [reply torch/csrc/jit/codegen/cuda/parser.cpp]: In this situation, we shouldn't see `train==False`.  The `aten::dropout` signature is where we get `train==False` and that is where tensor forwarding would apply.

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 9 above)
Comment (kevinstephano) [reply torch/csrc/jit/codegen/cuda/parser.cpp]: Added line back.

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 11 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp]: Oops, I wasn't careful enough ealier.

It is dangerous for us to assume that `dropout` is used for inference and `native_dropout` is used for training, even though this is totally the case if we use `torch.nn.Dropout`.

`torch.nn.functional.dropout` is exposed at user level. So it is totally possible for them to give arbitrary argument for `training` with combination to `requires_grad_` on input.

**Review by jjsjann123 — APPROVED** [review]
Awesome. Thanks for handling the comments
